### PR TITLE
Add support for extension frames.

### PR DESCRIPTION
**Pull request by Lukasa**
Builds on @haikuginger's work from #75.

### CODE REVIEW

**Comment by Lukasa** [comment]
Cool, so if we can get this landed we can move forward on adding support for unrecognised frames to h2, which is one of the big glaring omissions in our RFC 7540 support.

File: hyperframe/frame.py [hunk 1]
```diff
@@ -736,6 +745,61 @@ def parse_body(self, data):
...
+        # First, get the flags.
+        flags = self.flag_byte
+
+        header = _STRUCT_HBBBL.pack(
+            (self.body_len >> 8) & 0xFFFF,  # Length spread over top 24 bits
+            self.body_len & 0xFF,
+            self.type,
+            flags,
+            self.stream_id & 0x7FFFFFFF  # Stream ID is 32 bits.
+        )
```
Comment (alexwlchan) [inline hyperframe/frame.py]: I’m mildly uncomfortable about repeating what seems like a somewhat fiddly bit of code from the original `serialize()` method, but I’m not sure pulling it out into a function makes it better. 😕

Comment (Lukasa) [inline hyperframe/frame.py]: Yeah, this is a situation where a macro would be really nice. Curse you Python and your lack of macros. :frown:

**Comment by Lukasa** [comment]
So...why does Github think the statuses haven't all passed?

**Comment by alexwlchan** [comment]
> why does Github think the statuses haven't all passed?

Does for me? I just don’t have write access, apparently.

screen shot 2017-03-07 at 09 49 29

**Comment by Lukasa** [comment]
This is my current view:

screen shot 2017-03-07 at 09 49 44

I'll add the team with write access and we'll see whether you can merge.

**Comment by Lukasa** [comment]
Also, you totally do have write access: 

screen shot 2017-03-07 at 09 51 20

**Comment by Lukasa** [comment]
Oh wait, forgot that there are two settings that have to be used here. Try now?

**Comment by alexwlchan** [comment]
🤔

So I wonder if:

1. Either there’s a status check GitHub is expecting that isn’t showing up (although I can’t imagine what that is)
2. Restarting one of the Travis jobs would prod it into life?

**Comment by alexwlchan** [comment]
Did you actually change anything? The appearance has changed for me:

screen shot 2017-03-07 at 09 55 13

**Comment by Lukasa** [comment]
I fixed the permissions. I keep forgetting that with protected branches I need to give a team write access *twice*: once to the repo as a whole, and then once to the branch. It's a pretty silly system but there we are. So now you can see what I can see: a broken GitHub PR.

**Comment by Lukasa** [comment]
Ok I'm just going to override GitHub right now.

**Comment by haikuginger** [comment]
@Lukasa, thanks for finishing this off! Apologies for letting it wither on the vine a bit; life has been a bit hectic of late.

**Comment by Lukasa** [comment]
That's not a problem at all, @haikuginger: you should volunteer only as much time as you feel comfortable doing and not a second more. The work will get done eventually. :smile: Take whatever time for yourself you need: we'll be here.

**Comment by Lukasa** [comment]
@alexwlchan Ok, I've written a tool to automatically manage branch permissions because I'm sick of this going wrong. I've just run it against the various Python hyper repositories and it should have reconciled their protected branches so it should now be possible for y'all to push to them.

**Pull request merged**